### ISSUE TITLE
Expected single tone output power?

### ISSUE THREAD

**Issue by sr-cdf** [realtimeradio/souk-firmware#1]
Hey @jack-h ,

The datasheet DS926 for the ZU48DR says the max DAC output power should be in the range -18.5 to +6.5 dBm.

Using `gen_lut.set_output_freq()`, measured a single tone at a range of frequencies and it peaks around -12dBm, which is nicely in that range, but can we get it higher? Is there a DAC setting?

![screenshot_100MHz](https://user-images.githubusercontent.com/17500915/235957918-980df1fb-2f88-4f4e-810f-7f54adc775cc.png)

![single_tone_output_power_1](https://user-images.githubusercontent.com/17500915/235957866-482ac666-6477-4863-b33e-09170468c032.png)

**Comment by sr-cdf** [comment]
And for the record, the RF baluns provide 1 or 2 dB insertion loss:
![Screenshot_20230504_100833](https://user-images.githubusercontent.com/17500915/236160731-ecb1f324-a700-465f-af1d-97f484d61b4e.png)

**Comment by sr-cdf** [comment]
@jack-h @mitchburnett, apologies for the lack of brevity here...

I've been playing with the Pynq image on the RFSoC4x2 and I measure -3dBm out of DAC_A straight out of the box, and I can tweak it up to +3dBm :
```
$ ssh xilinx@rfsoc4x2
xilinx@pynq:~$ sudo bash
root@pynq:/home/xilinx# source  /etc/profile.d/xrt_setup.sh
root@pynq:/home/xilinx# source /usr/local/share/pynq-venv/bin/activate
(pynq-venv) root@pynq:/home/xilinx# ipython

Python 3.10.4 (main, Apr  2 2022, 09:04:19) [GCC 11.2.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from pynq.overlays.base import BaseOverlay

In [2]: base = BaseOverlay('base.bit')

In [3]: from rfsystem.spectrum_sweep import ToneGenerator

In [4]: generator=ToneGenerator(channel=base.radio.transmitter.channel[1])

In [5]: generator.frequency_selector.centre_frequency = 600 #MHz

In [6]: generator.amplitude_controller.enable = True

In [7]: generator.amplitude_controller.gain = 1.0
```

This gave -3.0dBm on my spectrum analyser. 

The fine mixer has an optional "3dBV" attenuator and toggling this drops the powers down by 1.25dB  to -4.25dBm:
```
In [8]: d20=base.radio.rfdc.dac_tiles[2].blocks[0]

In [9]: d20.MixerSettings
Out[9]: 
{'Freq': 600.0,       #float, fine mixer NCO freq
 'PhaseOffset': 0.0,  # float, NCO phase offset
 'EventSource': 0,    # XRFDC_EVNT_SRC_IMMEDIATE
 'CoarseMixFreq': 0,  # XRFDC_COARSE_MIX_OFF
 'MixerMode': 2,      # XRFDC_MIXER_MODE_C2R
 'FineMixerScale': 1, # XRFDC_MIXER_SCALE_1P0
 'MixerType': 2}      # XRFDC_MIXER_TYPE_FINE

In [10]: d20.MixerSettings['FineMixerScale']=2 #XRFDC_MIXER_SCALE_0P7, power drops by 1.25dB

In [11]: d20.MixerSettings['FineMixerScale']=1 # return to full power.

```

The quadrature modulation correction block has the option to add up to a factor of 2 in gain, and setting this raises the tone power slightly to -2.35dBm:

```
In [12]: d20.QMCSettings['GainCorrectionFactor']   #default value is 0
Out[12]: 0.0

In [13]: d20.QMCSettings['EnableGain']   #default setting is Off
Out[13]: 0

In [14]: d20.QMCSettings['EnableGain']=1 #Enabling with zero gain kills the tone

In [15]: d20.QMCSettings['GainCorrectionFactor']=1.0 # Setting gain revives the tone

In [16]: d20.QMCSettings['GainCorrectionFactor']=1.999 # doubling the gain boosts tone by 0.65dB

In [17]: d20.QMCSettings['GainCorrectionFactor']=1.0 # return to original tone

In [18]: d20.QMCSettings['EnableGain']=0 #disable QMC, tone remains.

```

Adjusting the Variable Output Power (only available with Gen3 chips) from 20mA (default current) up to the maximum 40.5 mA (expecting 6.12dB power rise) raises the tone power by 5.45dB up to +2.45dBm:

```
 In [19]: d20.OutputCurr #returns current in microAmps, default is 20mA
Out[19]: 19993

In [20]: d20.SetDACVOP(40500) #max allowed setting is 40.5mA, this gives +2.45dBm out

In [21]: d20.OutputCurr  #checking the new value, its close to the set value
Out[21]: 40468
```

Bearing in mind that this firmware is simply feeding a DC level (amplitude_controller.gain) to the DAC and the fine NCO mixer produces the tone at the NCO freq; what if we try with the coarse mixer as we do in our firmware?

```
In [23]: import xrfdc

In [24]: d20.MixerSettings={'CoarseMixFreq': xrfdc.COARSE_MIX_MIN_SAMPLE_FREQ_BY_FOUR,
    ...:                    'MixerMode':xrfdc.MIXER_MODE_C2R,
    ...:                    'MixerType':xrfdc.MIXER_TYPE_COARSE}
```
This gives +3.2dBm at 1228.8 MHz (Fs/4) for max output current (40.5mA), which is about as much as I can squeeze out of it for now.

So, a question is: is there a way a to adjust the various RFDC settings using the xrfdc driver from within the CASPER image?



Now if I return to our firmware on the CASPER image and try to repeat this, I notice something interesting...

Using the LUT generator, and requesting a sine wave at 1000 MHz (IF=-228.8 MHz + LO=Fs/4) with `r.gen_lut.set_output_freq(0,1000e6-r.adc_clk_hz/2)` gives a tone -10.9dBm. The LUT is composed of sine waves for I and Q, both ranging from -1 to 1, multiplied by 2**14, giving extrema at +/-16384 . This should probably be clipped to [-16384,+16383] to prevent overflow/wrapping.

Using the LUT generator at 0 Hz IF with `r.gen_lut.set_output_freq(0,0)` gives a tone at Fs/4 with -10.8dBm. I is all ones, Q is all zeros, the phase = 0.

Setting all of the LUT elements directly to 0.9999+0j  with `r.gen_lut.set_lut_output(0,zeros(r.gen_lut.n_samples,dtype=complex) + 0.9999*(1+0j) )` gives an identical tone at -10.8dBm.

Scaling this up by 1 bit to the full signed 16 bit range (double the magnitude) with `r.gen_lut.set_lut_output(0,zeros(r.gen_lut.n_samples,dtype=complex) + 2*0.9999*(1+0j) )` gives a tone at -5.8 dBm. This is only a 5dB increase, not the expected 6.

Then, if I set Q = I , effectively increasing the magnitude by sqrt(2) and rotating by 45deg in IQ plane, the tone power doubles up as expected to -2.5dBm, which is just about where we started with the Pynq image. But this power increase is only possible when the frequency is 0Hz and the phase = pi*n/4 (n=1,3,5,7)

If I could control the DAC Output Current with xrfdc, then I expect this could be pushed further up to +3dBm.

Trying again with the CORDICs, this is with v5.5 firmware with 4 cordics and an adder tree. The CORDIC output is forced to zero when the phase is set to zero, so I can't generate at exactly Fs/4, but setting the frequency to 1Hz, gives 1228.799999 MHz output which is close enough. And there is no amplitude control with the CORDICs, unless I set all 4 to the same phase (x4 V = +12dB). Lo and behold, the single tone power is -22.8dBm and when adding all 4, it's back up to -10.8dBm as it was with the LUT. So it seems there is 6dB to be gained here from a lost MSB somewhere. 

Similarly with the Polyphase Synthesis Bank, the single tone power (with no integer scaling) at Fs/4 is -10.8dBm, suggesting that there is one MSB missing in the pipeline somewhere as there was with the LUT. 

I have a strong feeling that finding this missing bit will increase the dynamic range by 6dB and enable upto -4dBm output power. 

And being able to tweaking the Variable Output Power setting from 20mA to 40mA would  boost the output power by a further 6dB, but I don't know how that would affect the SNR.

Do you guys agree?

**Comment by mitchburnett** [comment]
@sr-cdf, regarding capabilities of the current `rfdc` driver capabilities in `casperfpga`; With respect to the mixer, I suspect your `tcpborphserver3` (TBS) and `casperfpga`  versions only have the ability to set the NCO frequency and to report the mixer settings. Not very helpful, I know, sorry. To see the true list of functions you have, you could netcat or telnet to your board on port 7147 (e.g., `nc rfsoc4x2 7147`) and get a raw TBS  connection. Here, enter `?help` and TBS will return a list of functions. Look for functions prefixed `rfdc`. For example, `?rfdc-report-mixer 0 0 dac` will provide mixer settings for DAC tile 0, block 0.

I started a version of TBS that started supporting more of the mixer functionality, like setting that fine mixer scale setting (going from `MIXER_SCALE_1P0` to `MIXER_SCALE_0P7`). I believe I have only got to adding in for the ADCs and then became distracted with other tasks. Adding the same for the DACs is no problem. The situation is similar for the variable output power (VOP) setting. The related setting on the ADC side is the digital step attenuator (DSA). This is supported with the `get/set-dsa` commands (and their `casperfpga` equivalent). So, we would need to extend the driver to include the the ability to adjust the VOP at runtime. It is configurable when building the firmware in the `rfdc` yellow block mask. Adjustments with QMC also need to be added. Extending these capabilities in the `rfdc` for CASPER is straight forward. I would like to be a help to you here as much as I can. This looks like a great motivator to get this properly finished.

To reconcile the missing bit in the pipeline I would need a bit more clarification. I have gone through the LUT/polyphase synth firmware, and supporting software, to really only play with it. If a missing bit sparks anything for @jack-h, then. However, if there is some other discrepancy between firmware output and heading into the `rfdc` I would need a bit more help taking the results of your experiments here, and getting oriented with what is an apple-to-apple comparison in terms of cause and expected effect. At least in the firmware it looks consistent, I at least understood that. 🙃

**Comment by jack-h** [comment]
@sr-cdf thanks for these results. I think all your conclusions are correct.

Regarding the "missing bit" -- it's worth knowing that the firmware/software is generally constructed to make undetectable overflow impossible, which I think gives rise to what you see.
Three relevant facts:
1. The LUT generator (which stores signed 16-bit values for each of I and Q) scales so that saturation is at 2, not 1, as you saw. This is to prevent a unity-amplitude sine wave from overflowing. This behaviour could easily be changed if you want. Options might be:

**Comment by mitchburnett** [comment]
@jack-h you aren't misunderstanding. This is, right. VDAC_AVTT is set to 2.5V for the 4x2 and the PG has its recommendation for operating VOP. This is one of those things though where it still works, does not damage the hardware, but is now out of spec, and the performance outlined in DS926 is no longer guaranteed. I would also expect a slightly degraded SNR.

**Comment by sr-cdf** [comment]
Thanks both.

My main desire here is to make full use of the DAC and ADC ranges to maximise the SNR. Secondary to that is to utilise VOP and DSA to maximise/optimise the power out and in.

@jack-h I'd say aggressive is the way to go. Can you reconfigure all the generators to output at the full 16-bit signed range, and include some integer overflow or carry-flag-to-counter-to-snapblock logic at the appropriate places? 

I believe the LUT is already good and just needs a software tweak to scale by the extra 1 bit without wrapping at +32768.

Can the 4x or 1x CORDICs be set to 16-bit output each, and then added together with some overflow detection? Is it also possible to add a scaling factor to them as you did with the LO CORDIC??

And with the PSB, can you remove the division so a single signed 16 bit full-scale tone (or a combination of tones) can be fed to the DAC, and add some overflow logic to the adder?

I'd rather be able to generate a full scale signal and check for overflow, than have a guarantee of no overflow at the expense of dynamic range.

Considering that DAC VOP is not supported, despite it actually working, it would be nice to take some measurements of SNR as a function of the VOP settings but this would ideally require on-the-fly adjustment. The additional 5 or 6 dB to be gained might remove the need for amplification later in the RF system. Similarly so for the extra 6dB to be gained by using the full scale range of the DAC.


@mitchburnett thanks for the info and I look forward to using a fully functioning real-time driver in casperfpga in the not so distant future :) Is there a reason you are building your own, rather than installing/running the xrfdc driver and python bindings on the casper image?

I've had varied success using the `?rfdc-set-dsa [tile] [block] [atten_db]` command over telnet, some points to note are...
- the available settings don't quite match up with the data sheet spec and there is a strange initial setting, and the attenuation units should be in units of dB (power = 20 log(amplitude)) but it appears to be operating in units of 10 log(amplitude)...
  - `?rfdc-set-dsa 0 0 x` (RFSoC4x2 SMA connector 'ADC_D') command works fine for `0 <= x <= 11.0`, 
  - `?rfdc-set-dsa 0 0 1.0` attenuates the signal power recorded with the ADC Snapshot by 2.0 dB
  - `?rfdc-set-dsa 0 0 0.5` attenuates the signal power by 1.0 dB
  - fractional settings are always rounded up to the nearest 0.5 dB setting (1dB actual attenuation).
  - '?rfdc-get-dsa` always reports the DSA setting as integer despite the fractional settings working as above.
  - values less than zero fail with `min supported atten is 0.0 dB`
  - `?rfdc-get-dsa` always reports `DSA: -2 dB` after programming the board, and the actual attenuation does appear to be 4dB lower than with the 0 dB setting.
  - DS926 states the range is 0 to 27 dB in 1dB steps but as above I can only get `0 <= att <= 22.0`, or att = -4 dB (after programming)
- if the signal power into the ADC exceeds the over-voltage level (~1dBm), the RFSoC force sets the DSA to 15dB, measurements with ADC snapshot confirm this. 
  - `?rfdc-get-dsa` does not see this forced value and continues to report the latest values set with `?rfdc-set-dsa`, or the initial value of -2 dB. 
  - `?rfdc-set-dsa` is also oblivious to this and returns success but the level is never actually changed. 
  - reducing the input amplitude to below the limit does not de-assert the overflow and the RFSoC needs to be reprogrammed to reset the DSA. Re-initialising the RFDC in software does not clear the over-amplitude.

Its probably too much to ask for you to implement every possible setting of the driver! But at least for now it would be nice to have real time control of the VOP current, the full range of DSA settings and the ability to inspect/clear the various ADC overflow flags. Then lower priority would be the various settings for ADC Threshold detection, Sub-RF-ADC Cal , DAC/ADC QMC/Mixer,  DAC InvSinc etc.

Is it too much to ask?


```
$ telnet rfsoc4x2 7147
Trying 192.168.2.42...
Connected to rfsoc4x2.
Escape character is '^]'.
#version-connect katcp-library v0.2.0-291-g8b2d8aa-dirty 2022-05-26T22:43:15
#version-connect katcp-protocol 5.0-M
#version-connect kernel 5.4.0-xilinx-v2020.2 #1\_SMP\_Mon\_Aug\_15\_19:08:58\_UTC\_2022
#log info 1700492539.992 raw unmapping\_fpga
#fpga loaded
#fpga down
#log info 1700492540.024 raw awaiting\_transfer\_on\_port\_2316
#log info 1700492540.149 raw new\_client\_connection\_127.0.0.1:37234
#client-connected 127.0.0.1:37234
#fpga loaded
#log info 1700492541.931 raw mapping\_area\_0x000ad000\_at\_0xa0000000
#fpga mapped
#fpga ready
#log info 1700492540.151 kcpfpg retrying\_connect\_to\_port\_7146
#log info 1700492541.481 kcpfpg fpga\_programmed
#log info 1700492541.938 raw received\_end\_of\_file\_from\_127.0.0.1:37234
#client-disconnected 127.0.0.1:37234
#log info 1700492541.946 raw awaiting\_transfer\_on\_port\_2075
#log info 1700492542.048 raw overlay\_path\_already\_exists
#log info 1700492542.048 raw will\_apply\_overlay
#log info 1700492542.048 raw removing\_active\_overlay\_first
#log info 1700492542.050 raw creating\_dto\_casper
#log info 1700492542.051 raw writing\_dto
#log info 1700492552.043 raw unmapping\_fpga
#fpga loaded
#fpga down
#log info 1700492552.075 raw awaiting\_transfer\_on\_port\_2198
#log info 1700492552.201 raw new\_client\_connection\_127.0.0.1:37240
#client-connected 127.0.0.1:37240
#fpga loaded
#log info 1700492554.490 raw mapping\_area\_0x000ad000\_at\_0xa0000000
#fpga mapped
#fpga ready
#log info 1700492552.203 kcpfpg retrying\_connect\_to\_port\_7146
#log info 1700492554.044 kcpfpg fpga\_programmed
#log info 1700492554.496 raw received\_end\_of\_file\_from\_127.0.0.1:37240
#client-disconnected 127.0.0.1:37240
#log info 1700492554.519 raw awaiting\_transfer\_on\_port\_2426
#log info 1700492554.622 raw overlay\_path\_already\_exists
#log info 1700492554.622 raw will\_apply\_overlay
#log info 1700492554.622 raw removing\_active\_overlay\_first
#log info 1700492554.623 raw creating\_dto\_casper
#log info 1700492554.623 raw writing\_dto

?rfdc-get-dsa
#rfdc-get-dsa ADC:0,\_BLK:0,\_DSA:\_-2\_dB
#rfdc-get-dsa ADC:0,\_BLK:1,\_DSA:\_-2\_dB
#log info 1700493466.216 raw ADC:0,\_BLK:2\_(disabled)
#log info 1700493466.216 raw ADC:0,\_BLK:3\_(disabled)
#log info 1700493466.216 raw ADC:1\_(disabled)
#log info 1700493466.216 raw ADC:2\_(disabled)
#log info 1700493466.216 raw ADC:3\_(disabled)
!rfdc-get-dsa ok

?rfdc-set-dsa 0 0 0
#log info 1700493593.259 raw request\_set\_dsa\_for\_tile:\_0,\_blk:\_0\_to\_0.0\_dB
#rfdc-set-dsa ADC:0,\_BLK:0,\_DSA:\_0\_dB
!rfdc-set-dsa ok

?rfdc-get-dsa
#rfdc-get-dsa ADC:0,\_BLK:0,\_DSA:\_0\_dB
#rfdc-get-dsa ADC:0,\_BLK:1,\_DSA:\_-2\_dB
#log info 1700493599.869 raw ADC:0,\_BLK:2\_(disabled)
#log info 1700493599.869 raw ADC:0,\_BLK:3\_(disabled)
#log info 1700493599.869 raw ADC:1\_(disabled)
#log info 1700493599.869 raw ADC:2\_(disabled)
#log info 1700493599.869 raw ADC:3\_(disabled)
!rfdc-get-dsa ok

?rfdc-set-dsa 0 0 -2
#log error 1700493627.601 raw invalid\_attenuation\_setting\_--2.0,\_min\_supported\_atten\_is\_0.0\_dB
!rfdc-set-dsa invalid

?rfdc-set-dsa 0 0 10    
#log info 1700493695.706 raw request\_set\_dsa\_for\_tile:\_0,\_blk:\_0\_to\_10.0\_dB
#rfdc-set-dsa ADC:0,\_BLK:0,\_DSA:\_10\_dB
!rfdc-set-dsa ok

?rfdc-set-dsa 0 0 11
#log info 1700493702.329 raw request\_set\_dsa\_for\_tile:\_0,\_blk:\_0\_to\_11.0\_dB
#rfdc-set-dsa ADC:0,\_BLK:0,\_DSA:\_11\_dB
!rfdc-set-dsa ok

?rfdc-set-dsa 0 0 12
#log info 1700493705.289 raw request\_set\_dsa\_for\_tile:\_0,\_blk:\_0\_to\_12.0\_dB
#log error 1700493705.289 raw failure\_to\_set\_dsa
!rfdc-set-dsa fail
```

**Comment by jack-h** [comment]
@sr-cdf -- I'll try and implement these features tomorrow on the v5.5 designs (which I assume you are still using)
@mitchburnett -- if you do feel the urge to work on some of the features mentioned, I have money to pay for this. Gimme a shout on slack.

**Comment by mitchburnett** [comment]
@sr-cdf - Depending on which version of `casperfpga` ends up being the dependency will determine how much of the upstream python bindings exist. For example, it looks like in this version you do have the [`get/set_dsa()`](https://github.com/realtimeradio/casperfpga/blob/974b2ea8b96f2417c2cb6a326b6fb6b7afc65fcf/src/rfdc.py#L332) methods in `casperfpga` that you can use rather than going directly through katcp. The motivation for directing you to the `telnet` side of things was because this is how you would always tell what for sure would be in the image you have.

Tcpborphserver wraps the call to the lower-level `xrfdc` driver. Then, `casperfpga` provides a python binding to making a convenient client interface to work with the platform. So, it is not much that this is reimplementation of existing driver. Rather, the lower-level `xrfdc` driver provides some methods and so there is some conditioning and setup that needs to be done prior to making the call.

For example, I expect much of what you are experiencing with the `get_set/dsa()` involves a nuanced mix of which version of the underlying `xrfdc` driver version and that when these functions were being implemented it was targeting my ZCU216 on the ALPACA instrument where those are ES silicon, and the 4x2 is not. The difference being that different steps and resolution are allowed based on ES and non-ES parts. The CASPER version was trying to be friendly to all flavors of target RFSoC, and obviously has some bugs. Thanks for testing and detailing these out.

I'll reach out and work something out with @jack-h on moving forward with this. Fixing DSA and supporting VOP is straightforward. I vaguely remembering doing some testing with the DSA and needing to do some reconciliation with the PG/DS about why the over-protection levels are not reflected in the DSA. What you are reporting about needing to reset the over-protection is true. This I a little more involved because there is some underlying HDL work that needs to be done to properly reset this. But, I could be remembering wrong. I always need to jump back into the PG/DS and then some things start coming back to me. This was all to say that supporting all of these need to happen. The priority and level-of-effort varies based on what infrastructure exists to do that.

**Comment by jack-h** [comment]
@sr-cdf -- Looking at the model now, and I think have made the necessary tweaks to get full scale from the LUT and CORDIC generators.
Looking at the synthesis bank(s), I actually think my claim was wrong that the odd and even are divided by 2 before being added together, so I'm not quite sure why you see 6dB lower than expected. That said, the following seems to be true:
1. The input to the PSB cannot be set to full scale (for the same reason that the CORDIC generators add a bit to set the max range to +/-2)
2. The FIR front-end of the PSB should, I think, add about 2dB of loss to the synth bank paths.

I could change the PSB inputs so that the CORDICs generate full scale, but my guess is that all this will really achieve is that you'll need to add another shift stage to avoid overflow, and the output will remain the same.
I suspect that maybe with multiple PSB tones you can recover the full range, but maybe a more general purpose solution is to have another scaling stage after the PSB -- this means that:
1. you can compensate for any FIR loss
2. having set a shift schedule, you can always get full DAC range, even if the number of tones means that the shift required to avoid saturation doesn't allow full scale PSB output.

That sound reasonable?

**Comment by sr-cdf** [comment]
@jack-h That sounds good to me.

It makes sense that scaling up the CORDIC prior to the FFT might cause the FFT to overflow and then require downshifting. 

Previously with one tone, FFT saturation was not possible. With two tones, there was saturation, so shifting at the first stage was required. So, as you say, scaling up the CORDIC will simply saturate the FFT and it will then need shifting down.

But the single tone output power from the DAC was identical to the case of the LUT using 15 bits not 16. Which would suggest to me the PSB output simply needs to be multiplied by 2 to reach the DAC full scale. Although, this doesn't account for any loss in the FIR, so I'm not 100% sure.

Adding another scaling stage at the PSB output, with an overflow flag attached, seems like the way forward.

I think I've now come to realise that even though the DACs/ADCs have 14 bits, all of the 16 bits on the RFDC interface are useful. The top 14 bits should contain the signal of interest, but the internal mixers and other components will end up putting/taking information to/from the last 2 bits. [https://support.xilinx.com/s/question/0D52E00006iHmqTSAS/rfsoc-adc-output-data-format-with-mixer-and-decimation?language=en_US](https://support.xilinx.com/s/question/0D52E00006iHmqTSAS/rfsoc-adc-output-data-format-with-mixer-and-decimation?language=en_US)

@mitchburnett I'm glad you are onboard and thanks for explaining how things work with xrfdc and casperfpga. If the driver is installed on the casper image, then it it may be useful for me (in the short term) to use it directly without going through casperfpga, since we are running code on the RFSoC ARM anyway. Is the source code for tcpborphserver hosted somewhere public so I can see how you interact with the xrfdc driver? 

For what it's worth, I tried using the DSA with the python pynq tools on the Pynq Image, and it also did not notice the over-amplitude protection kicking in, and also required a reload of the overlay. I think I'll try to find out more over on the Pynq support forum.

In any case, please let me know if there is anything I can do to help with testing anywhere along the way.

Cheers

**Comment by jack-h** [comment]
@sr-cdf -- compiling now. Will update when I've checked it does something reasonable.

I also don't quite understand why the PSB would be exactly the same power as the LUT. I'll poke around a bit when I have the new firmware.
Yes, all 16-bits are useful -- sorry, I could have highlighted this earlier.

I think the relevant tcpborphserver code is at https://github.com/realtimeradio/katcp_devel/blob/rfsoc/tcpborphserver3/rfsoc.c
E.g. https://github.com/realtimeradio/katcp_devel/blob/85c11dc3e7b93bca04929e42a406416dc99e87bd/tcpborphserver3/rfsoc.c#L1074
The underlying driver code is at https://github.com/Xilinx/embeddedsw/blob/master/XilinxProcessorIPLib/drivers/rfdc/src/xrfdc.h (though tcpborphserver will rely on a particular branch -- probably easiest to look on the SD card image and see which version is checked out). If this is critical and not easy to find I can probably dig the info out.

I'd assume that there is a linux device driver available via sysfs entries, so that you can just `cat` some values to set things like the drive strength. I can't immediately find any documentation to help here. Again, I can get on this if you'd like, but for now I'm just worrying about firmware.
Maybe I'm being naive, but I'd assume that increasing the drive strength doesn't improve the SNR (and maybe makes it worse if operating out of spec), but this doesn't take into account the benefits of potentially removing some amps

**Comment by mitchburnett** [comment]
@sr-cdf, the TBS code is also [hosted here](https://gitlab.ras.byu.edu/alpaca/casper/katcp/-/tree/rfsoc/rfdc). This will be where I end up working from. What @jack-h pointed to will be the version of TBS that would have been compiled to replace what is default shipped with the CASPER RFSoC4x2 image.

To say "[the] the driver is installed on the casper image" needs to be clarified. The `xrfdc` driver is a shared object library, `librfdc.so`, compiled from a version of the source that @jack-h linked from Xilinx embedded software drivers. By default, that library lives at `/opt/local/lib/` on the CASPER RFSoC images. TBS is compiled against this library to exercise the functional capabilities of the driver (so more like an API). Unfortunately, there is no `sysfs` bindings for the driver that exposes these functions. If there exists an implementation out there, that would be an interesting capability to consider.

You could run `casperfpga` locally on the rfsoc4x2? But, this is just a suggestion, I am not familiar with your various requirements. I myself will skip `casperfpga` for directly poking TBS connecting locally from the RFSoC4x2, just for testing.

Let me know if you find anything interesting with the PYNQ forum. Although, in my experience, because some of this is still new, much is loosely documented, trial and error, and many implementations face the same issues. For example, PYNQ does the same thing by exercising the `librfdc` library by loading that library. Said differently, since you reported similar behavior, exercising and interfacing correctly with the over-protection capabilities in the RFDC will be through different mechanisms not associated with the DSA.

I do think increasing the drive strength will degrade SNR rather than improve it. The output SNR will be affected by the noise figure of the device increasing the DAC out and corrupt the input SNR. Unfortunately, these types of numbers are not provided readily by RFSoC data sheets and need to be requested of Xilinx by FAEs. So, although it may hurt SNR, any testing you want to do to report on this based on empirical tests I do think are valuable.

**Comment by mitchburnett** [comment]
@sr-cdf, earlier, when discussing "ADC overflow" you are indicating the ability to detect the over voltage protection capabilities of the rfdc, correct? This would be in contrast to something like ADC FIFO overflows.

**Comment by jack-h** [comment]
@mitchburnett -- Yeah, this is all power level related
@sr-cdf - See https://github.com/realtimeradio/souk-firmware/commit/6e96a35432a02f939c2fb7047fd506e87f1a5e29

I have:

1. Added amplitude setting on the output CORDIC . The 4 CORDICs sum without scaling. A new `gen_cordic.get_cordic_overflows` has been added to detect overflows.
2. Defaulted the LUT output to 95% full scale (`0.95 * (1 - 1/2**gen_lut._n_bp)`). You can override to a bit higher, but my experience was that close to full scale, even though the DAC input doesn't saturate, the output as viewed on a spectrum analyzer has spurs.
3. Added scaling and overflow detection after summing the two PSBs, prior to casting to 16-bits. See the new block object `SoukMkidReadout.psbscale` and `SoukMkidReadout.set_output_psb_scale(nshift, scale=1., check_overflow=True)`. For a single tone, with nshift=0, and scale=2, you should get something close to full scale. The scale value is, internally, 16 bits with 8 bits below the binary point.

I haven't, but could, set the output power at build time to the out-of-spec 40mA setting, but I figured I'd wait for you to learn more with the pynq image before doing that. FWIW, I've committed the latest .bit and .hwh files, in case that helps use pynq with the CASPER build.

**Comment by sr-cdf** [comment]
@jack-h I'll take a look tomorrow and log some signal and noise measurements to confirm dynamic range or enobs or whatever it is. Then I'll compare with what I get with the default pynq overlay. 

@mitchburnett, as Jack said, it's over-voltage not over-flow, apologies for the imprecise language. 

I also agree that setting a higher output current will degrade the signal/noise but I'd like to see by how much and in what way. If the effective noise figure is low, it might be worth using the extra gain if the analog system needs it.

And thanks for pointing me to the TBS code and clarifying the setup. 

How hard would it be to implement a filesystem based approach to setting the driver parameters? Are we talking a one-off late night or more like a couple of long weeks?

It would also be cool if we could drop Pynq on top of the CASPER image and use the Pynq xrfdc module alongside the python casperfpga. I'll let you know if I have any luck with that...

**Comment by jack-h** [comment]
I spoke to @mitchburnett yesterday and we converged on a plan to try and satisfy your "at least for now it would be nice to have real time control of the VOP current, the full range of DSA settings" which seem very useful for testing.
I'm not sure if we're meeting on Thurs, but I can give you the long version of that conversation, and the various difficulty steps in enabling some of the other functionality.

I think implementing some `sysfs` thing would be quite a project. If it's worth doing at all (and my suspicion is that it isn't), it's probably worth doing in communication with the Xilinx PYNQ team (or someone else from Xilinx device driver world) who may already have some starting points / may know where the main difficulties are.

**Comment by mitchburnett** [comment]
@jack-h, @sr-cdf -- I have some updates for you to test.

I have an [ftp site](http://casper.groups.et.byu.net/souk-updates/) where you can go and download a new version of `tcpborphserver` and the supporting rfdc library, `librfdc`. There is an accompanying readme for steps on how to install and replace these, if you are not already familiar with how to do so. These updates correctly supports the DSA range and settings, and can now get the output current and set the VOP current. There is a slight change from the previous `get_dsa()`. Now, `get_dsa()` works on the tile and block level rather than returning all of the information in one call.

Let me know how these work for you. I did some preliminary testing on a device. But, only looking for software responses. Nothing looking at signals on a scope, or in a loopback.

There is a [casperfpga branch](https://github.com/mitchburnett/casperfpga/tree/rfsocs/rfdc-dsa-vop) with new and changed functions to support the changes. I suspect there will need to be some work in merging this branch into the fork of `casperfpga` for the firmware and some reconciling of the software. Because, this branch includes some updates to other functions like the `status()` function that now more conveniently parses the response and returns it as a dictionary rather than the previously useless output. I can raise a PR against your casperfpga? Or we can discuss for a bit how that should be handled?

**Comment by jack-h** [comment]
Thanks, @mitchburnett . I'll have a look at this today.
I don't think the merging should be too troublesome - I'll give it a go and see where things land.

**Comment by jack-h** [comment]
Also, can you point me to the source repo for borphserver @mitchburnett

**Comment by jack-h** [comment]
@sr-cdf I've merged the casperfpga changes into the submodule at https://github.com/realtimeradio/souk-firmware/tree/v5.6/software (i.e., v5.6 branch of the main repo).

After installing the new tcpborphserver and rfdc library per http://casper.groups.et.byu.net/souk-updates/readme things seem to be working as expected. I.e., I can read and set VOP and DSA parameters both from telnet and via casperfpga.
Via casperfpga, note that the `rfdc` object referred to in the readme is exposed as `SoukMkidReadout.rfdc.core` -- i.e., to set the attenuator for ADC_D on the 4x2 to 5dB -- `SoukMkidReadout.rfdc.core.set_dsa(0,0,5)`

One thing I notice is that when I crank up the output current I see the tone power go up as expected but I also see significant spurs, so I'm not sure this is a go-er. I'm not sure if this is an artefact of the DAC_AVTT not being 3V as the data sheet demands. (It is 2.5V on both the 4x2 and KRM boards)

Quick testing suggests the

**Comment by jack-h** [comment]
@sr-cdf noted that the above link is dead. I have submoduled the tcpborphserver source code in the main repo under `software/zynq/katcp_devel`

**Comment by sr-cdf** [comment]
I have not exhaustively tested all of @mitchburnett 's features but the VOP and DSA settings appear to work nicely :)

Some harmonic distortion comes in when VOP is >20000 uA, but this is expected given that the VDAC_AVTT is set to 2.5V.

**Comment by jack-h** [comment]
I think this is all done now and can be closed? @sr-cdf

**Comment by sr-cdf** [comment]
Yep

**Comment by sr-cdf** [comment]
I think it all makes sense :)
